[MLIR][Presburger] Use Identifiers outside Presburger library

iambrj opened the pull request:
The pull request includes the following changes.
1. Refactors the interface to `PresburgerSpace::identifiers` to `setId` and a
const `getId`, instead of previous `getId` which returned a mutable reference.
`resetIds` does not need to be called to use identifiers, `setId` calls
`resetIds` if identifiers are not enabled.
2. Deprecates `FlatAffineRelation` by refactoring all usages of
`FlatAffineRelation` to `IntegerRelation`. To achieve this,
`FlatAffineRelation::compose` is refactored into
`IntegerRelation::mergeAndCompose`.
3. Deletes unneeded overrides of virtual functions `hasConsistentState`,
`clearAndCopyFrom` and `fourierMotzkinEliminate` from
`FlatLinearValueConstraints` as these were only used through
`FlatAffineRelation` and we now use `IntegerRelation`'s member functions
instead.
4. Fixes an existing bug in FlatLinearValueConstraints' constructor which caused
identifiers set by superclass FlatLinearConstraints' constructor to be erased.
5. Fixes `IntegerRelation::convertVarKind` not preserving identifiers.

iambrj committed:
[MLIR][Presburger] Implement preserve identifiers in IntegerRelation::convertVarKind

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -217,11 +219,11 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
-      values.resize(getNumDimAndSymbolVars(), std::nullopt);
-    else
-      values.append(valArgs.begin(), valArgs.end());
+    // Store Values in space's identifiers.
+    space.resetIds();
+    for (unsigned i = 0, e = valArgs.size(); i < e; ++i)
+      if (valArgs[i])
+        setValue(i, *valArgs[i]);
```
It should be documented somewhere how the SSA Values of the constraint system are stored (as Identifiers in the Space)

iambrj replied:
added comment in FlatLinearValueConstraints' constructor [here](https://github.com/llvm/llvm-project/pull/77316/files#diff-e77afb47e1a45788a8eb369ae11eb5303a9d0c554b7686bb7baf1dae1f5372beR208)

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -272,11 +274,15 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
     if (valArgs.empty())
-      values.resize(getNumDimAndSymbolVars(), std::nullopt);
-    else
-      values.append(valArgs.begin(), valArgs.end());
+      return;
+    assert(valArgs.size() == getNumDimAndSymbolVars());
+    // Store Values in space's identifiers.
+    space.resetIds();
+    for (unsigned i = 0, e = valArgs.size(); i < e; ++i)
+      if (valArgs[i])
+        setValue(i, *valArgs[i]);
```
i think instead of doing resetIds everywhere, maybe we want to hide this. Can we add a check in PresburgerSpace that if ids are not being used and someone wants to use them, enable them, otherwise do not populate the Identifiers vector.

iambrj replied:
added comment in FlatLinearValueConstraints' constructor [here](https://github.com/llvm/llvm-project/pull/77316/files#diff-e77afb47e1a45788a8eb369ae11eb5303a9d0c554b7686bb7baf1dae1f5372beR208)

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -313,27 +318,47 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
+  inline SmallVector<std::optional<Value>> getMaybeValues() const {
+    SmallVector<std::optional<Value>> maybeValues;
+    maybeValues.reserve(getNumDimAndSymbolVars());
+    for (unsigned i = 0, e = getNumDimAndSymbolVars(); i < e; ++i)
+      if (hasValue(i))
+        maybeValues.push_back(getValue(i));
+      else
+        maybeValues.push_back(std::nullopt);
```
nit: use braces

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -313,27 +318,47 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
+    return space.getId(kind, relativePos).hasValue();
   }
 
+  void resetValues() { space.resetIds(); }
```
This should be part of IntegerRelation?

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -444,28 +468,6 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
   LogicalResult unionBoundingBox(const FlatLinearValueConstraints &other);
   using IntegerPolyhedron::unionBoundingBox;
-
-protected:
-  /// Eliminates the variable at the specified position using Fourier-Motzkin
-  /// variable elimination, but uses Gaussian elimination if there is an
-  /// equality involving that variable. If the result of the elimination is
-  /// integer exact, `*isResultIntegerExact` is set to true. If `darkShadow` is
-  /// set to true, a potential under approximation (subset) of the rational
-  /// shadow / exact integer shadow is computed.
-  // See implementation comments for more details.
-  void fourierMotzkinEliminate(unsigned pos, bool darkShadow = false,
-                               bool *isResultIntegerExact = nullptr) override;
-
-  /// Returns false if the fields corresponding to various variable counts, or
-  /// equality/inequality buffer sizes aren't consistent; true otherwise. This
-  /// is meant to be used within an assert internally.
-  bool hasConsistentState() const override;
-
-  /// Values corresponding to the (column) non-local variables of this
-  /// constraint system appearing in the order the variables correspond to
-  /// columns. Variables that aren't associated with any Value are set to
-  /// std::nullopt.
-  SmallVector<std::optional<Value>, 8> values;
```
nice, we can remove virtual classes from Presburger soon

Groverkss commented on mlir/include/mlir/Analysis/Presburger/IntegerRelation.h:
```diff
@@ -674,6 +674,11 @@ class IntegerRelation {
+  /// Given a relation `other: (A -> B)`, this operation merges the symbol and
+  /// local variables and then takes the composition of `other` on `this: (B ->
+  /// C)`. The resulting relation represents tuples of the form: `A -> C`.
+  void mergeAndCompose(const IntegerRelation &other);
```
We should make this default behavior of compose honestly. Not for today though.

Groverkss commented on mlir/lib/Dialect/Affine/Analysis/AffineAnalysis.cpp:
```diff
@@ -411,7 +412,7 @@ addOrderingConstraints(const FlatAffineValueConstraints &srcDomain,
     const FlatAffineValueConstraints &srcDomain,
     const FlatAffineValueConstraints &dstDomain, unsigned loopDepth,
-    FlatAffineValueConstraints *dependenceDomain,
+    IntegerRelation *dependenceDomain,
```
Why is domain a relation? Shouldnt this be a set?

Groverkss commented on mlir/lib/Dialect/Affine/Analysis/AffineAnalysis.cpp:
```diff
@@ -457,7 +459,7 @@ static void computeDirectionVector(
   }
 }
 
-LogicalResult MemRefAccess::getAccessRelation(FlatAffineRelation &rel) const {
+LogicalResult MemRefAccess::getAccessRelation(IntegerRelation &rel) const {
```
We should be able to remove FlatAffineRelation in this next patch. Nice!

Groverkss reviewed (changes requested):
Please add a description of the patch. My only major comment is that we should hide resetIds and instead PresburgerSpace should automatically enable ids if someone is trying to use them.

Groverkss commented on mlir/include/mlir/Analysis/FlatLinearValueConstraints.h:
```diff
@@ -272,11 +272,13 @@ class FlatLinearValueConstraints : public FlatLinearConstraints {
                              ArrayRef<std::optional<Value>> valArgs = {})
       : FlatLinearConstraints(fac) {
-    assert(valArgs.empty() || valArgs.size() == getNumDimAndSymbolVars());
+    // Do not reset values assigned by FlatLinearConstraints' constructor.
```
Redundant comment?

Groverkss commented on mlir/include/mlir/Analysis/Presburger/PresburgerSpace.h:
```diff
@@ -250,21 +250,31 @@ class PresburgerSpace {
 
   ArrayRef<Identifier> getIds(VarKind kind) const {
     assert(kind != VarKind::Local && "Local variables have no identifiers");
+    assert(usingIds && "Identifiers not enabled for space");
```
this is not the best, but ok for now...

Groverkss commented on mlir/include/mlir/Analysis/Presburger/PresburgerSpace.h:
```diff
@@ -250,21 +250,31 @@ class PresburgerSpace {
     return {identifiers.data() + getVarKindOffset(kind), getNumVarKind(kind)};
   }
 
+  ArrayRef<Identifier> getIds() const { return identifiers; }
```
No check here for usingIds?

Groverkss reviewed (approved):
LGTM with some minor comments. This would be great to land! 🥳

iambrj committed:
[MLIR][Presburger] Use Identifiers outside Presburger library

Changes:
1. Refactors the interface to `PresburgerSpace::identifiers` to `setId`
   and a const `getId`, instead of previous `getId` which returned a
   mutable reference.  `resetIds` does not need to be called to use
   identifiers, `setId` calls `resetIds` if identifiers are not enabled.
2. Deprecates `FlatAffineRelation` by refactoring all usages of
   `FlatAffineRelation` to `IntegerRelation`. To achieve this,
   `FlatAffineRelation::compose` is refactored into
   `IntegerRelation::mergeAndCompose`.
3. Deletes unneeded overrides of virtual functions `hasConsistentState`,
   `clearAndCopyFrom` and `fourierMotzkinEliminate` from
   `FlatLinearValueConstraints` as these were only used through
   `FlatAffineRelation` and we now use `IntegerRelation`'s member
   functions instead.
4. Fixes an existing bug in FlatLinearValueConstraints' constructor
   which caused identifiers set by superclass FlatLinearConstraints'
   constructor to be erased.
